### PR TITLE
Handle concurrent open cursor requests

### PR DESCRIPTION
**Pull request by sboobna**
### Motivation

If the Managed Ledger gets concurrent open cursor requests for the same cursor, 1 request fails with Cursor already exists exception.

### Modifications

Any cursor that is uninitialized will be reused for the next open cursor request using a Completable Future for the cursor, so that it doesn't try to create a new cursor.

### Result

Managed Ledger will be able to handle concurrent open cursor requests properly.

### CODE REVIEW

**Comment by yahoocla** [comment]
CLA is valid!

**Comment by yahoocla** [comment]
CLA is valid!

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java [hunk 1]
```diff
@@ -554,21 +570,30 @@ public synchronized void asyncOpenCursor(final String cursorName, final OpenCurs
...
             @Override
             public void operationComplete() {
                 log.info("[{}] Opened new cursor: {}", name, cursor);
                 cursor.setActive();
-                cursors.add(cursor);
-
                 // Update the ack position (ignoring entries that were written while the cursor was being created)
                 cursor.initializeCursorPosition(getLastPositionAndCounter());
+
+                synchronized (this) {
```
Comment (rdhabalia) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java]: Now, as we are sure that only one thread calls ```cursor.initialize(..)``` then should we need ```synchronized``` here?

Comment (sboobna) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java]: But another thread will be accessing ```uninitializedCursors```, so we will need it to be synchronized

**Pull request merged**